tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -323,11 +323,138 @@ bool promotionImprovesCoalescing(
+ *
+ * Within each instance of the scope loops, check that loops that are either
+ * unrolled or mapped to threads access a single tensor element in the group
+ * (other loop indices will then not appear in the subscripts, making register
+ * promotion possible).  In other words, check that the relation between the
+ * flat product of prefix, thread-mapped, and unrolled loop indices and
+ * accessed elements is single-valued.
+ * TODO: note that if a group is formed from partially overlapping references,
+ * one must consider per-reference access relation for single-valuedness as
+ * different references may have different values, but all of them remain
+ * independent of non-unrolled loop iterators.
+ */
+bool accessSubscriptsAreUnrolledLoops(
+    const TensorReferenceGroup& group,
+    const detail::ScheduleTree* root,
+    const detail::ScheduleTree* scope,
+    isl::multi_union_pw_aff prefixSchedule,
+    isl::multi_union_pw_aff threadSchedule) {
+  using namespace detail;
+
+  auto nodes = ScheduleTree::collect(scope);
+  auto leaves = functional::Filter(
+      [](const ScheduleTree* tree) { return tree->numChildren() == 0; }, nodes);
+
+  auto domainNode = root->elemAs<detail::ScheduleTreeElemDomain>();
+  CHECK(domainNode);
+  auto domain = domainNode->domain_;
+
+  // Descend into every leaf.
+  for (auto leaf : leaves) {
+    auto ancestors = leaf->ancestors(root);
+    ancestors.push_back(leaf);
+    auto subdomain = activeDomainPointsBelow(root, leaf);
+
+    auto unrolledDims = isl::union_pw_aff_list(leaf->ctx_, 1);
+    for (auto node : ancestors) {
+      auto band = node->elemAs<detail::ScheduleTreeElemBand>();
+      if (!band) {
+        continue;
+      }
+
+      isl::multi_union_pw_aff schedule = band->mupa_;
+      if (isThreadMappedBand(node)) {
+        // Band members are mapped to threads in inverse order.  There may be
+        // less members than thread dimensions (the remaining thread dimensions
+        // are mapped to 0).  Transform the thread schedule accordingly before
+        // subtracting it from the band schedule.
+        auto bandCompatibleThreadSchedule =
+            isl::union_pw_aff_list(schedule.get_ctx(), band->nMember());
+        auto threadScheduleDims =
+            threadSchedule.get_union_pw_aff_list().reverse();
+        CHECK_LE(band->nMember(), static_cast<size_t>(threadScheduleDims.n()));
+        for (int i = 0; i < band->nMember(); ++i) {
+          bandCompatibleThreadSchedule =
+              bandCompatibleThreadSchedule.add(threadScheduleDims.get(i));
+        }
+
+        schedule = schedule.sub(isl::multi_union_pw_aff(
+            schedule.get_space(), bandCompatibleThreadSchedule));
+      }
+      schedule = schedule.intersect_domain(subdomain);
+
+      for (size_t i = 0, e = band->nMember(); i < e; ++i) {
+        if (!band->unroll_[i]) {
+          continue;
+        }
+        unrolledDims = unrolledDims.add(schedule.get_union_pw_aff(i));
+      }
+    }
```

Comment:
I would put the collection of the unrolled schedule into a separate function.
`accessSubscriptsAreUnrolledLoops` is pretty long now.

Reply:
not anymore

Reply:
>
> not anymore

I think it's still pretty long and it would be easier
to read if this was abstracted out.  I'll leave it like this for now.